server/datastore/mysql/operating_systems.go

```diff
@@ -51,8 +52,11 @@ func getOrGenerateOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hos
 // `newOperatingSystemDB` inserts a record for the given operating system and
 // returns the record including the newly associated ID.
 func newOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.OperatingSystem) (*fleet.OperatingSystem, error) {
-	stmt := "INSERT IGNORE INTO operating_systems (name, version, arch, kernel_version, platform) VALUES (?, ?, ?, ?, ?)"
+	stmt := "INSERT INTO operating_systems (name, version, arch, kernel_version, platform) VALUES (?, ?, ?, ?, ?)"
 	if _, err := tx.ExecContext(ctx, stmt, hostOS.Name, hostOS.Version, hostOS.Arch, hostOS.KernelVersion, hostOS.Platform); err != nil {
+		if strings.Contains(err.Error(), "Error 1062") {
+			return nil, doRetryErr
+		}
```

Comment:
two questions:

1. there's a comment in line `63` below about using `INSERT IGNORE` here, should we update some of that logic?
2. Why `doRetryErr`?

Reply:
Thanks @roperzh!

1. Yes, I think we can delete that and just use the last insert id instead.
2. Assuming you're referring to the retry line 58, we're returning the `doRetryErr` so that it bubbles back up to `UpdateHostOperatingSystems`. We're trying to cover the case where there is a race with two hosts trying to introduce the same operating system. In the majority of cases with `UpdateHostOperatingSystems` is getting called, it should be returning after the `getOperatingSystemDB` call with no attempt to insert. The problem I think we ran into here is in cases like right after this migration where numerous existing hosts are coming online and trying to report the same os. Definitely an added layer of complexity with the async approaches we've been trying to implement in the interest of minimizing writes.